### PR TITLE
Robert/16 session deauthentication

### PR DESCRIPTION
**Pull request by RobertSaylesAND**
[Ticket](https://github.com/isaaccomputerscience/isaac-cs-issues/issues/16) - ensure session (token) is de-authenticated on logout.

Update logout process to invalidate the session on logout by regenerating the associated token stored in the database.
This process was already employed as part of 'logout everywhere' and has been extended to standard logouts as well.
Updated session tokens to use randomised numbers instead of sequential integers.
Applied new session invalidation during rejection of invalid requests in filter.
Tidied up some logout-related error and response message strings.

### CODE REVIEW

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java [hunk 1]
```diff
@@ -498,24 +499,41 @@ public RegisteredUser createIncompleteLoginUserSession(final HttpServletRequest
...
+     * @throws NoUserLoggedInException - if a user cannot be retrieved from the session information
+     * @throws SegueDatabaseException  - if accessing the database fails
+     */
+    public void invalidateSessionToken(final HttpServletRequest request) throws NoUserLoggedInException, SegueDatabaseException {
+        Validate.notNull(request);
+        RegisteredUser currentUser = this.getUserFromSession(request, false);
+        if (null == currentUser) {
+            throw new NoUserLoggedInException();
+        }
+        this.database.regenerateSessionToken(currentUser);
```
Comment (charlescolbourn) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java:534]: It's not obvious what's going on here - it looks like the method is doing something completely different to it's name

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java [hunk 2]
```diff
@@ -1209,14 +1239,7 @@ public Cookie createAuthLogoutCookie() {
...
     public NewCookie createAuthLogoutNewCookie() {
-        NewCookie logoutCookie = new NewCookie.Builder(SEGUE_AUTH_COOKIE)
-                .value("")
-                .path("/")
-                .maxAge(0)
-                .httpOnly(true)
-                .secure(setSecureCookies)
-                .comment(SAME_SITE_LAX_COMMENT)
-                .build();
+        NewCookie logoutCookie = new NewCookie.Builder(SEGUE_AUTH_COOKIE).value("").path("/").maxAge(0).httpOnly(true).secure(setSecureCookies).comment(SAME_SITE_LAX_COMMENT).build();
```
Comment (charlescolbourn) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java:1242]: this was clearer before I think?

File: src/main/java/uk/ac/cam/cl/dtg/segue/dao/users/IUserDataManager.java [hunk 3]
```diff
@@ -15,14 +15,14 @@
...
 
-import uk.ac.cam.cl.dtg.segue.api.Constants.SchoolInfoStatus;
-import uk.ac.cam.cl.dtg.segue.api.Constants.TimeInterval;
-import uk.ac.cam.cl.dtg.segue.auth.AuthenticationProvider;
-import uk.ac.cam.cl.dtg.segue.dao.SegueDatabaseException;
 import uk.ac.cam.cl.dtg.isaac.dos.users.Gender;
 import uk.ac.cam.cl.dtg.isaac.dos.users.RegisteredUser;
 import uk.ac.cam.cl.dtg.isaac.dos.users.Role;
 import uk.ac.cam.cl.dtg.isaac.dos.users.UserAuthenticationSettings;
+import uk.ac.cam.cl.dtg.segue.api.Constants.SchoolInfoStatus;
```
Comment (charlescolbourn) [inline src/main/java/uk/ac/cam/cl/dtg/segue/dao/users/IUserDataManager.java:22]: why the change in import order?

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java (hunk 2 above)
Comment (RobertSaylesAND) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java:1242]: Hmm... must have been caught in a code format without me noticing. Ah, just after merging in the latest changes from main, I think. Agreed on the clarity, I'll put that one back...

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java (hunk 1 above)
Comment (charlescolbourn) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java:534]: where is `regenerateSessionToken`? I can't find a method with that name in the isaac-api code

Comment (RobertSaylesAND) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java:534]: It's in the `pgUsers` class, replacing `incrementSessionToken` since we're not doing sequential tokens any more.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java [hunk 4]
```diff
@@ -498,24 +499,41 @@ public RegisteredUser createIncompleteLoginUserSession(final HttpServletRequest
...
+     * @param request - a servlet request holding an auth cookie for the user session to be invalidated
+     * @throws NoUserLoggedInException - if a user cannot be retrieved from the session information
+     * @throws SegueDatabaseException  - if accessing the database fails
+     */
+    public void invalidateSessionToken(final HttpServletRequest request) throws NoUserLoggedInException, SegueDatabaseException {
+        Validate.notNull(request);
+        RegisteredUser currentUser = this.getUserFromSession(request, false);
+        if (null == currentUser) {
+            throw new NoUserLoggedInException();
+        }
```
Comment (charlescolbourn) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java:533]: `//by regenerating the token without updating it in the cookie we stop them matching and invalidate the session`

File: src/main/java/uk/ac/cam/cl/dtg/segue/dao/users/PgUsers.java [hunk 5]
```diff
@@ -666,14 +669,36 @@ public void updateUserLastSeen(final RegisteredUser user, final Date date) throw
...
+        Integer newSessionToken = generateRandomTokenInteger();
+        this.updateSessionToken(user, newSessionToken);
+        return newSessionToken;
+    }
+
+    private static Integer generateRandomTokenInteger() {
+        Random random = new SecureRandom();
+        int newValue = random.nextInt();
+        // -1 is reserved for 'no assigned token', used for when a user is logged out for example
+        if (newValue != -1) return newValue;
```
Comment (charlescolbourn) [inline src/main/java/uk/ac/cam/cl/dtg/segue/dao/users/PgUsers.java:682]: This /could/ return -1. It's not hugely likely, but it could. I suppose it could also return 0, which we also don't want?

Comment (RobertSaylesAND) [reply src/main/java/uk/ac/cam/cl/dtg/segue/dao/users/PgUsers.java:682]: Is there something I'm missing here? It should only return a value that passes the check of not being -1, right? Otherwise it recursively calls the function for a different value, which is also subject to the check of not equalling -1. Unless we hit the infinitesimally small probability of it getting -1 every time and recursing until it runs out of memory or something, I don't think we should have an issue here?
0 isn't actually a problem in and of itself. The problem before was that the initial session token for _every_ user was _always_ 0, and further it would remain 0 until they used the `logout everywhere` option instead of the standard `logout` option.

**Review by charlescolbourn — APPROVED** [review]
